### PR TITLE
allow user to specify custom version id for flex deployments

### LINKED ISSUE #572: Allow users to enter custom versions and module name for flex deployments

### PR DESCRIPTION
**Pull request by etanshaul**
fixes #572 

Version override feature is implemented:

image

### CODE REVIEW

**Comment by etanshaul** [comment]
Just realized, I need to implement the placeholder text feature (something like 'auto' when there is no version override).

**Comment by patflynn** [comment]
yeah I would put something like 'automatically generated', since we have
lots of horizontal space to work with.

File: google-cloud-tools-plugin/src/com/google/gct/idea/appengine/cloud/AppEngineDeploymentRunConfigurationEditor.java [hunk 1]
```diff
@@ -187,6 +199,7 @@ protected void applyEditorTo(AppEngineDeploymentConfiguration configuration)
     configuration.setDockerFilePath(dockerFilePathField.getText());
     configuration.setAppYamlPath(appYamlPathField.getText());
     configuration.setConfigType(getConfigType());
+    configuration.setVersion(versionOverrideCheckBox.isSelected() ? versionIdField.getText() : null);
```
Comment (chanseokoh) [inline google-cloud-tools-plugin/src/com/google/gct/idea/appengine/cloud/AppEngineDeploymentRunConfigurationEditor.java]: Should there be validation to exclude, e.g., white spaces, special characters?

Comment (etanshaul) [inline google-cloud-tools-plugin/src/com/google/gct/idea/appengine/cloud/AppEngineDeploymentRunConfigurationEditor.java]: yes, I'll add a regex. The pattern is printed out from gcloud when an invalid version string is sent.

**Comment by etanshaul** [comment]
Updated with placeholder text (I was working on creating a reusable JTextField subclass that adds placeholder functionality. Putting a "hardcoded" version out for review in the meantime):

image

Regarding @chanseokoh's point about validation: wouldn't it be best to just let it pass through to gcloud and let it do the validation? This minimizes client logic that we need to maintain (if the validation pattern is ever changed for instance).

**Comment by chanseokoh** [comment]
About validation: that's a nice idea. That said, I just notice that --version is wrapped in an unusual way in the actual command:

--quiet "--version=some name" --format=json app.yaml

which results in bizarre error message. It will be better to be --version="some name". I guess it's doing automatic escaping then?

**Comment by etanshaul** [comment]
Updated placeholder logic to use a barebones JTextField subclass.

**Comment by etanshaul** [comment]
@chanseokoh what error are you seeing? The placement of the version id hasn't changed. Could it be that you were using an invalid version pattern? I don't think whitespace is allowed.

**Comment by chanseokoh** [comment]
I just tried an unacceptable name with a space for the sake of testing. The message is

Fetching storage object: gs://staging.springboot-maven-project.appspot.com/us.gcr.io/springboot-maven-project/default.some name:latest#1459799851639000
CommandException: No URLs matched: gs://staging.springboot-maven-project.appspot.com/us.gcr.io/springboot-maven-project/default.some
ERROR
ERROR: fetching gcs source: running gsutil: exit status 1
ERROR: (gcloud.preview.app.deploy) Error Response: [2] Build failed; check build logs for details

**Comment by etanshaul** [comment]
Ah. I guess its GeneralCommandLine that's doing that escaping. If you try with another invalid version id, e.g. MyId (caps aren't allowed), you'll get a better error message. I'll think about a solution - ideally we wouldn't duplicate the validation that gcloud already does.

**Comment by chanseokoh** [comment]
I think we are fine with white space. I tested with gcloud with --version="some name" and it returned the same weird message. We don't want to duplicate validation.

**Comment by etanshaul** [comment]
Updated:
- fixed bug discussed offline with @chanseokoh where serialized state was not being restored
- added version validation: if override is selected but version is empty, it warns the user (and prevents deploy)

@chanseokoh @patflynn ptal

File: google-cloud-tools-plugin/src/com/google/gct/idea/appengine/cloud/AppEngineDeploymentRunConfigurationEditor.java [hunk 2]
```diff
@@ -168,15 +177,22 @@ public File get() {
                 .defaultDockerfile(DeploymentArtifactType.typeForPath(deploymentSource.getFile()));
           }
         }, dockerFilePathField, userSpecifiedArtifactFileSelector));
+    versionOverrideCheckBox.addItemListener(
+        new CustomFieldOverrideListener( versionOverrideCheckBox, versionIdField));
```
Comment (patflynn) [inline google-cloud-tools-plugin/src/com/google/gct/idea/appengine/cloud/AppEngineDeploymentRunConfigurationEditor.java]: there's an errant white space before versionOverrideCheckBox

**Comment by patflynn** [comment]
LGTM

**Comment by chanseokoh** [comment]
The bug is gone. LGTM.

**Pull request merged**